### PR TITLE
virtual-libjpeg: initial commit

### PR DESCRIPTION
**Pull request by ledocc**
Specify library name and version:  **virtual-libjpeg/9.0**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

add virtual package for jpeg implementation. cf https://github.com/conan-io/conan-center-index/issues/2894

### CODE REVIEW

**Comment by ledocc** [comment]
conan-center hook throw many error, but a virtual package
- have not licenses or homepage
- create empty package
- have no settings
- have no source to download, so no conandata.yml is need

any chance to disable some test for virtual package ?

File: recipes/virtual-libjpeg/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+class VirtualjpegConan(ConanFile):
+    name = "virtual-libjpeg"
+    description = "Virtual package to choose jpeg encoder/decoder implementation in final package."
+    url = "https://github.com/conan-io/conan-center-index"
+    options = {"implementation": "ANY"}
+    default_options = {"implementation": "libjpeg-turbo/2.0.5"}
+
+    def requirements(self):
+        self.requires(str(self.options.implementation))
```
Comment (madebr) [inline recipes/virtual-libjpeg/all/conanfile.py:12]: - Not a fan of putting the name+version as a value.
This is not an easy format to parse by consumers.
- What default should we choose?
I though most recipes used `libjpeg` by default and `libjpeg-turbo` is patched in.
- Also, should the CI be modified to check, when a recipe requires this virtual package, whether the recipe actually is able to build with both libjpeg implementations?
- Is there a better sounding synomym for *implementation*?
```suggestion
    options = {
        "implementation": ["libjpeg", "libjpeg-turbo"],
    }
    default_options = {
        "implementation": "libjpeg-turbo",
    }

    def requirements(self):
        if self.options.implementation == "libjpeg":
            self.requires("libjpeg/9d")
        elif self.options.implementation == "libjpeg-turbo":
            self.requires("libjpeg-turbo/2.0.5")
        self.requires(str(self.options.implementation))
```

File: recipes/virtual-libjpeg/config.yml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+versions:
+    "9.0":
```
Comment (madebr) [inline recipes/virtual-libjpeg/config.yml:2]:
```suggestion
    "virtual":
```

File: recipes/virtual-libjpeg/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+from conans import ConanFile, tools
+
+
+class VirtualjpegConan(ConanFile):
+    name = "virtual-libjpeg"
```
Comment (Croydon) [inline recipes/virtual-libjpeg/all/conanfile.py:5]: I think a better name would be just `jpeg`

File: recipes/virtual-libjpeg/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    description = "Virtual package to choose jpeg encoder/decoder implementation in final package."
+    url = "https://github.com/conan-io/conan-center-index"
+
+    options = {
+        "implementation": ["libjpeg", "libjpeg-turbo"],
```
Comment (Croydon) [inline recipes/virtual-libjpeg/all/conanfile.py:10]: When I first suggested a virtual package for OpenGL, I used the term `provider` instead, which I would personally like better than `implementation`.

What do you all think?

File: recipes/virtual-libjpeg/all/conanfile.py (hunk 3 above)
Comment (madebr) [reply recipes/virtual-libjpeg/all/conanfile.py:5]: We have:
- openjpeg
- libjpeg
- libjpeg-turboe
- mozjpeg
- openjpeg

I think jpeg is too generic.

**Comment by SpaceIm** [comment]
mozjpeg also provides jpeg API (and also libjpeg-turbo API, therefore a libjpeg-turbo proxy package should also be created).

**Comment by SpaceIm** [comment]
What do you think of adding JPEG for cmake names in package info, so that we can consume any libjpeg implementation for free in CMake based libraries?

File: recipes/virtual-libjpeg/all/conanfile.py (hunk 3 above)
Comment (Croydon) [reply recipes/virtual-libjpeg/all/conanfile.py:5]: * mozjpeg: It's compatible with libjpeg API and ABI, and can be used as a drop-in replacement for libjpeg https://github.com/mozilla/mozjpeg#mozilla-jpeg-encoder-project-

If mozjpeg is another drop in replacement for libjpeg und libjpeg-turbo it should be added as an option too

Openjpeg seems to be it's own thing indeed (two times in your list btw)

I guess we can argue about `libjpeg` or `jpeg`, maybe `libjpeg` is actually better, but I think it is weird to have `virtual-` in it's name. We don't add `system-` to the name for system packages either

Comment (ledocc) [reply recipes/virtual-libjpeg/all/conanfile.py:5]: "Virtual package" should be used by package creator over "real package". 
Adding "virtual-" in front of the name allow to quickly find virtual package by doing `conan search virtual-*`.

File: recipes/virtual-libjpeg/all/conanfile.py (hunk 4 above)
Comment (ledocc) [reply recipes/virtual-libjpeg/all/conanfile.py]: good, and that match to "provides" attribute introduce in conan 1.29.0

File: recipes/virtual-libjpeg/all/conanfile.py (hunk 3 above)
Comment (ledocc) [reply recipes/virtual-libjpeg/all/conanfile.py:5]: openjpeg and mozjpeg added

File: recipes/virtual-libjpeg/config.yml (hunk 2 above)
Comment (ledocc) [reply recipes/virtual-libjpeg/config.yml:2]: jpeg interface have many version 7.0, 8.0 and 9.0 (the last one) and we provide a virtual package for the version 9. I think we should use the version of the API that this virtual package provide.

File: recipes/virtual-libjpeg/all/conanfile.py (hunk 3 above)
Comment (SpaceIm) [reply recipes/virtual-libjpeg/all/conanfile.py:5]: openjpeg doesn't provide libjpeg API

File: recipes/virtual-libjpeg/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
+from conans import ConanFile, tools
+
+
+class VirtualjpegConan(ConanFile):
+    name = "virtual-libjpeg"
+    description = "Virtual package to choose jpeg encoder/decoder implementation in final package."
+    url = "https://github.com/conan-io/conan-center-index"
+
+    options = {
```
Comment (SpaceIm) [inline recipes/virtual-libjpeg/all/conanfile.py:9]: options should be deleted in package_id

File: recipes/virtual-libjpeg/config.yml (hunk 2 above)
Comment (SpaceIm) [reply recipes/virtual-libjpeg/config.yml:2]: So we should also provide a mapping between libjpeg 9 API and libjpeg-turbo and mozjpeg compatible versions.

**Comment by jgsogo** [comment]
Hi! We need to move forward with this PR. Yesterday, @SSE4 proposed a different approach (with the same capabilities). I would like to know what you think about it:

```python

class Recipe(ConanFile):

    options = {"provider": "ANY"}
    default_options = {"provider": "libjpeg/9d"}  # Default to the latest version of the preferred alternative

    def requirements(self):
        self.requires(self.options.provider)

    ...
```

Using it from the CLI:

```
conan install -o libjpeg-virtual:provider=libjpeg-turbo/2.0.5
```

from `conanfile.txt`:

```ini
[requires]
lib1/version
lib2/version

[options]
libjpeg-virtual:provider=libjpeg-turbo/2.0.5
```

---

What do you think? It has one main advantage and it is that only one explicit library is hardcoded in the recipe, it should be the _preferred_ one and we need only to modify the recipe when a new version of this library appears. It is also easily extensible if any new recipe provides _libjpeg_ API, no need to modify the _virtual_ recipe, everything is on the consumer side. 

OTH, the drawback is that the alternatives are not listed and there is no information about which libraries are interchangeable here...  maybe we could add some validation:

```python
name, _ = self.options.provider.split('/', 1)
if name not in ['libjpeg', 'libjpeg-turbo',...]:
   raise ConanInvalidConfiguration(...)
```

It is still better over listing these references with versions...

**Comment by jgsogo** [comment]
I see something very similar was initially proposed in the PR, I feel the same as @madebr  in one of the first reviews

> Not a fan of putting the name+version as a value.

...but the more I think about it, and about the pros and cons and alternatives, it is an option I find quite easy and maintainable.

**Comment by SpaceIm** [comment]
https://github.com/conan-io/conan/issues/12374

**Pull request closed without merging**